docs/building.md

```diff
@@ -0,0 +1,249 @@
+_muen_ target (which is not self-hosting), you should build Solo5 and
+unikernels on a system matching the host you will be running them on.
+
+## System requirements
+
+Supported _targets_, host operating systems/hypervisors and processor
+architectures:
+
+Production:
+
+* _ukvm_: Linux/KVM, using ukvm as a monitor, on the x86\_64 architecture.
+* _ukvm_; FreeBSD vmm, using ukvm as a monitor, on the x86\_64 architecture.
+  FreeBSD 11-RELEASE or later is required.
+
+Experimental:
+
+* _ukvm_: Linux/KVM, using ukvm as a monitor, on the aarch64 architecture. You
+  will need hardware capable of running a recent (v4.15+) 64-bit mainline
+  kernel and a 64-bit Linux distribution.
+* _ukvm_: OpenBSD vmm, using ukvm as a monitor, on the x86\_64 architecture.
+  OpenBSD 6.3 is known to work, however for correct operation with downstream
+  unikernel build systems, as of June 2018 OpenBSD-current is required.
+* _muen_: The Muen Separation Kernel, on the x86\_64 architecture. Plese see
+  [XXX](xxx) for Muen-specific instructions.
+
+Limited:
+
+* _virtio_: Any hypervisor which virtualizes an x86\_64 system with virtio
+  devices (e.g.  Linux/KVM with QEMU as a monitor, Google Compute Engine). See
+  below under "_virtio_: Limitations" for details.
+
+# Running Solo5-based unikernels
+
+Solo5 itself does not provide a high-level mangement or orchestration layer for
+unikernels -- this is intended to be provided by separate downstream projects.
+If you are coming from Linux containers you can think of Solo5 as conceptually
+occupying the same space in the stack as, for example, `runc`.
+
+If you are looking for a high-level stack for deploying unikernels, one such
+project is [Albatross](https://hannes.nqsb.io/Posts/VMM).
+
+The following examples use the standalone
+[test_ping_serve](tests/test_ping_serve/test_ping_serve.c) unikernel which is
+built as part of the normal Solo5 build process.
+
+The products of building a Solo5 unikernel are, depending on the _target_, one
+or two artifacts:
+
+- an ELF binary containing the built unikernel,
+- for the _ukvm_ target, a specialized "monitor", by convention currently built
+  as `ukvm-bin` alongside the unikernel binary.
+
+`test_ping_serve` is a minimalist network test which will respond only to ARP
+and ICMP echo requests sent to the hard-coded IP address of `10.0.0.2`. It
+accepts two possible command line arguments: Either `verbose` for verbose
+operation or `limit` to terminate after having sent 100,000 ICMP echo replies.
+
+## Setting up
+
+By convention, we will use the `tap100` interface to talk to the unikernel.
+
+To set up the `tap100` interface on Linux, run (as root):
+
+    ip tuntap add tap100 mode tap
+    ip addr add 10.0.0.1/24 dev tap100
+    ip link set dev tap100 up
+
+To set up vmm and the `tap100` interface on FreeBSD, run (as root):
+
+    kldload vmm
+    kldload if_tap
+    sysctl -w net.link.tap.up_on_open=1
+    ifconfig tap100 create 10.0.0.1/24 link0 up
```

Comment:
given the above `sysctl`, no need for the `up` -- and for `link0` I don't understand what that is supposed to do (`man ifconfig` tells me `link[0-2]` are device-specific, and I should look for further details in the device manpage, but `man tap` doesn't tell me any details).

Reply:
Did you want setup instructions for OpenBSD?

To set up vmm and the `tap100` interface on OpenBSD, run (as root):

    cd /dev
    ./MAKEDEV tap100
    ifconfig tap100 inet 10.0.0.1 netmask 255.255.255.0

Reply:
@hannesm You're right about `up`, but `link0` is required, otherwise when a ukvm instance is re-started the interface no longer gets packets routed to it correctly (I just re-tested this on `11.1-RELEASE`). 

The documentation is a bit confusing, but `tap(4)` says:

     On the last close of the data device, the interface is brought down (as
     if with ``ifconfig tapN down'') unless the device is a VMnet device, or
     has IFF_LINK0 flag set.  All queued frames are thrown away.  If the
     interface is up when the data device is not open, output frames are
     thrown away rather than letting them pile up.

Reply:
@adamsteen Thanks, will include those.